// expresswebcorets/lib/Validator/Manager.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const validatorjs_1 = __importDefault(require("validatorjs"));
class Manager {
    validateData(record, rules, customMessages = {}) {
        return new Promise((resolve, reject) => {
            this.validatorHelper(record, rules, customMessages, (err, result) => {
                if (err) {
                    reject({
                        success: false,
                        message: "Validation failed",
                        data: err,
                    });
                }
                if (result) {
                    resolve({
                        success: true,
                        message: "Validation passed",
                        data: record,
                    });
                }
            });
        });
    }
    validatorHelper(body, rules, customMessages, callback) {
        const validation = new validatorjs_1.default(body, rules, customMessages);
        validation.passes(() => callback(null, true));
        validation.fails(() => callback(validation.errors, false));
    }
}
exports.default = Manager;
